cdap-app-fabric/src/main/java/co/cask/cdap/route/store/ZKRouteStore.java

```diff
@@ -154,13 +155,48 @@ public void onSuccess(NodeData result) {
 
       @Override
       public void onFailure(Throwable t) {
-        routeConfigMap.remove(serviceId);
         settableFuture.setException(t);
+        // If no node is present, then simply place a watch on the node to monitor further data updates
+        // If its an exception other than NoNodeException, then remove the future, so that zkClient#getData
+        // will be retried during the next fetch config request
+        if (t instanceof KeeperException.NoNodeException) {
+          existsAndWatch(serviceId);
+        } else {
+          routeConfigMap.remove(serviceId);
```

Comment:
I have made ZKRouteStore#fetch return null in case of any error during the fetch in the latest commit. That way, in the UserEndpointRoutingStrategy, we will use the default routing strategy.